crates/re_ui/src/modal.rs

```diff
@@ -0,0 +1,184 @@
+///
+/// # egui::__run_test_ui(|ui| {
+/// #   let re_ui = ReUi::load_and_apply(ui.ctx());
+/// #   let re_ui = &re_ui;
+/// if ui.button("Open").clicked() {
+///     modal_handler.open();
+/// }
+///
+/// modal_handler.ui(re_ui, ui, || Modal::new("Modal Window"), |_, ui| {
+///     ui.label("Modal content");
+/// });
+/// # });
+/// ```
+#[derive(Default)]
+pub struct ModalHandler {
+    modal: Option<Modal>,
+    should_open: bool,
+}
+
+impl ModalHandler {
+    /// Open the model next time the [`ModalHandler::ui`] method is called.
+    pub fn open(&mut self) {
+        self.should_open = true;
+    }
+
+    /// Draw the modal window, creating/destroying it as required.
+    pub fn ui<R>(
+        &mut self,
+        re_ui: &crate::ReUi,
+        ui: &mut egui::Ui,
+        make_modal: impl FnOnce() -> Modal,
+        content_ui: impl FnOnce(&crate::ReUi, &mut egui::Ui) -> R,
+    ) -> Option<R> {
+        if self.modal.is_none() && self.should_open {
+            self.modal = Some(make_modal());
+            self.should_open = false;
+        }
+
+        if let Some(modal) = &mut self.modal {
+            let ModalResponse { inner, open } = modal.ui(re_ui, ui, content_ui);
+
+            if !open {
+                self.modal = None;
+            }
+
+            inner
+        } else {
+            None
+        }
+    }
+}
+
+/// Response returned by [`Modal::ui`].
+pub struct ModalResponse<R> {
+    /// What the content closure returned, if it was actually run.
+    pub inner: Option<R>,
+
+    /// Whether the modal should remain open.
+    pub open: bool,
+}
+
+/// Show a modal window with Rerun style.
+///
+/// [`Modal`] fakes as a modal window, since egui [doesn't have them yet](https://github.com/emilk/egui/issues/686).
+/// This is typically use via the [`ModalHandler`] helper object.
+pub struct Modal {
+    title: String,
+    default_height: Option<f32>,
+}
+
+impl Modal {
+    /// Create a new modal with the given title.
+    pub fn new(title: &str) -> Self {
+        Self {
+            title: title.to_owned(),
+            default_height: None,
+        }
+    }
+
+    /// Set the default height of the modal window.
+    #[inline]
+    pub fn default_height(mut self, default_height: f32) -> Self {
+        self.default_height = Some(default_height);
+        self
+    }
+
+    /// Show the modal window.
+    ///
+    /// Typically called by [`ModalHandler::ui`].
+    pub fn ui<R>(
+        &mut self,
+        re_ui: &crate::ReUi,
+        ui: &mut egui::Ui,
+        content_ui: impl FnOnce(&crate::ReUi, &mut egui::Ui) -> R,
+    ) -> ModalResponse<R> {
+        Self::dim_background(ui);
+
+        let mut open = ui.input(|i| !i.key_pressed(egui::Key::Escape));
+
+        let mut window = egui::Window::new(&self.title)
+            .pivot(egui::Align2::CENTER_CENTER)
+            .fixed_pos(ui.ctx().screen_rect().center())
+            .collapsible(false)
+            .resizable(true)
```

Comment:
Try adding `.contrain_to(ui.screen_rect())`. I think that should fix the centering.